warehouse/forklift/legacy.py

```diff
@@ -607,6 +609,24 @@ def file_upload(request):
             )
         raise _exc_with_message(HTTPForbidden, msg)
 
+    # If this is a user identity (i.e: API token) but there exists
+    # a trusted publisher for this project, send an email warning that an
+    # API token was used to upload a project where Trusted Publishing is configured.
+    # Only do this once per (API token, project) combination.
+    if request.user and project.oidc_publishers:
+        macaroon: Macaroon = request.identity.macaroon
+        # If we haven't warned about use of this particular API token and project
+        # combination, send the warning email
+        if macaroon not in project.macaroons_tp_warning:
+            send_api_token_used_in_trusted_publisher_project_email(
+                request,
+                project.users,
+                project_name=project.name,
+                token_owner_username=request.user.username,
+                token_name=macaroon.description,
+            )
+            project.macaroons_tp_warning.append(macaroon)
```

Comment:
We'd need to do this too:
```suggestion
            request.db.add(
                ProjectMacaroonWarningAssociation(
                    macaroon_id=macaroon.id,
                    project_id=project.id,
                )
            )
```

Reply:
done!